Vernus Andrew: RefTest for UserService.AddUser() method

Nerewan opened the pull request:
A little bit changes in naming
 + Separate some methods

Nerewan committed:
A little bit changes for naming
 + Separate some methods

L1feIsGood commented on LegacyApp/UserService.cs:
```diff
@@ -4,23 +4,9 @@ namespace LegacyApp
     public class UserService
     {
-        public bool AddUser(string firName, string surname, string email, DateTime dateOfBirth, int clientId)
+        public bool AddUser(string firstName, string surname, string email, DateTime dateOfBirth, int clientId)
```
Изменив интерфейс метода вы поломали совместимость для вызывающего кода, что не укладывается в рамки рефакторинга.
Следующий вызов в Program.cs перестанет компилироваться, что противоречит условию задачи о его неизменяемости: _userService.AddUser(firName: "Andrew", surname: "Vernus", email: ...

L1feIsGood commented on LegacyApp/UserService.cs:
```diff
@@ -33,45 +19,93 @@ public bool AddUser(string firName, string surname, string email, DateTime dateO
+            }
 
             return true;
         }
```
Жаль, что остановились только на Extract method. В идеале было бы еще избавить класс от этой ответственности.

L1feIsGood commented on LegacyApp/UserService.cs:
```diff
@@ -33,45 +19,93 @@ public bool AddUser(string firName, string surname, string email, DateTime dateO
+            }
+        }
+
+        #endregion
```
А здесь, помимо первых двух шагов, стоило подумать над тем, как изменить код так, чтобы при добавлении нового типа клиента не пришлось изменять уже написанный и работающий код (например появится RoyalClient, у которого помимо множителя лимита будет минимальный безусловный лимит)

L1feIsGood reviewed (changes requested):
В целом неплохо. Откровенно неудачных решений не было, но и удачные до конца довести также не удалось. Если есть желание, можно доработать.

Nerewan committed:
Extract validation methods to independent services
Rollback naming changes from previous commit
Add default value for User.HasCreditLimit = true

L1feIsGood commented on LegacyApp/CreditLimitsServiceClient.cs:
```diff
@@ -0,0 +1,48 @@
+        private readonly int mMinCreditLimit = 500;
+
+
+        public static ICreditLimitsService Create()
```
Какой-то неуклюжий синглтон получился. Что мне мешает в вызывающем коде написать 
var creditLimitsServiceClient = new CreditLimitsServiceClient()?
А, учитывая, что теперь даже в консольном приложении рекомендуется использовать DI Container с уже реализованным lifetime контролем, самостоятельная реализация синглтона выглядит архаизмом...

Nerewan replied:
Это не для синглтона, а для заглушки в конструкторе UserService'а (тот, что без параметров), т.к. по условию задачи нельзя менять код в Program.cs, но и вызывать конструктор напрямую из целевого кода не хотелось
Фактически, я просто не стал выделять какую-то отдельную фабрику для сервиса, а обошёлся полумерами, частично отвязав зависимость UserService от конструктора других сервисов

L1feIsGood commented on LegacyApp/CreditLimitsServiceClient.cs:
```diff
@@ -0,0 +1,48 @@
+            return new CreditLimitsServiceClient();
+        }
+
+        public ICreditLimitsService SetCreditLimit(User user)
+        {
+            var creditMultiplier = mDefaultCreditMultiplier;
+
+            switch(user.Client.Name)
+            {
+                case "VeryImportantClient":
+                    user.HasCreditLimit = false;
+                    break;
+                case "ImportantClilent":
+                    creditMultiplier = 2;
+                    break;
+                default:
+                    creditMultiplier = mDefaultCreditMultiplier;
+                    break;
+            }
+
+            if(user.HasCreditLimit)
+            {
+                using(var userCreditService = new UserCreditServiceClient())
+                {
+                    var creditLimit = userCreditService.GetCreditLimit(user.FirstName, user.Surname, user.DateOfBirth);
+                    user.CreditLimit = creditLimit * creditMultiplier;
+                }
+            }
+
+            return this;
+        }
```
Последняя проблема с расширением типов клиентов так и осталась нерешенной, к сожалению, но это задание со звездочкой, так что сойдет и так.
Немного непонятен return this, как будто требуется какой-то method chaining

Nerewan replied:
А можно намёк, в какую сторону подумать для решения подобной проблемы? Мне так, для себя, на будущее
Спасибо

L1feIsGood replied:
За решение был бы принят вариант с отдельным классом-наследником ICreditLimitsService (VeryImportantClientCreditLimitsService, ImportantClilentCreditLimitsService, ...) для каждого типа клиента, полностью инкапсулирующий логику расчета лимита и какая-либо фабрика, выбирающая на основе имени клиента конкретную реализацию и инстанцирующая ее для каждого запроса. 
Однако даже такой вариант не полностью избавляет от необходимости вносить правки в switch оператор в фабрике при выборе реализации, но хотя бы избавляет от проблем пересечения логики расчета в данном решении. Поэтому это вполне приемлемое решение.

При этом существует способ полностью избавиться от switch-а и просто добавлять новые классы, реализующие ICreditLimitsService, для новых типов клиентов. Можно динамически подтягивать список классов, реализующих ICreditLimitsService, из данных сборки (typeof(ICreditLimitsService).Assembly.ExportedTypes...) и инстанцировать их (с помощью Activator.CreateInstance), а подходящую реализацию выбирать с помощью linq. Но так как многие разработчики на дух не переносят рефлексию, а кто-то жалуется на ее быстродействие, то этот способ необязателен.

А если бы кому-нибудь удалось его реализовать, мы бы сильно удивились и обрадовались)

L1feIsGood commented on LegacyApp/ICreditLimitsService.cs:
```diff
@@ -0,0 +1,8 @@
+    public interface ICreditLimitsService
+    {
+        ICreditLimitsService SetCreditLimit(User user);
+        bool ValidateCreditLimit(User user);
```
Немного неудачный нейминг, кажется, что
bool IsCreditLimitValid(User user)
лучше описывает то, что делает метод. (как собственно и Check... из прошлого коммита)

L1feIsGood commented on LegacyApp/User.cs:
```diff
@@ -10,6 +10,6 @@ public class User
         public string Surname { get; set; }
         public int CreditLimit { get; set; }
-        public bool HasCreditLimit { get; set; }
+        public bool HasCreditLimit { get; set; } = true;
```
А это почему так? Вдруг в базе default false?

Nerewan replied:
Хм, справедливо. С одной стороны.
С другой стороны, по оригинальному коду можно сделать довольно вероятное предположение, что бизнес-правило для массового "юзера" - быть ограниченным по лимиту, и лишь для особых клиентов выдаются послабления в виде его отключения
Но это домыслы, а не чёткая инфа из ТЗ, так что да, с этой стороны - косяк...

L1feIsGood commented on LegacyApp/UserService.cs:
```diff
@@ -4,108 +4,63 @@ namespace LegacyApp
+            //Need more DI for the DI god!1 
+            mPersonalDataValidationService = PersonalDataValidationServiceClient.Create();
+            mCreditLimitsService = CreditLimitsServiceClient.Create();
         }
 
-        private bool CheckName(string firstName, string surname)
+        public UserService(IPersonalDataValidationService dataValidationService, ICreditLimitsService creditLimitsService)
         {
-            if(string.IsNullOrEmpty(firstName) || string.IsNullOrEmpty(surname))
-            {
-                return false;
-            }
-
-            return true;
+            mPersonalDataValidationService = dataValidationService;
+            mCreditLimitsService = creditLimitsService;
```
Какая-то каша получилась совсем...
1. Почему конструктор без параметров вдруг устарел? Он всего лишь использует зависимости по умолчанию
2. Плохая практика несколько раз писать реализацию конструктора, можно что-то забыть или не учесть при изменении. В данном случае гораздо лучше было бы (не учитывая избыточных синглтонов)
public UserService() : this (PersonalDataValidationServiceClient.Create(), CreditLimitsServiceClient.Create()) {}

Nerewan replied:
1. Для обозначения того, что в будущем лучше использовать новый конструктор, но старый мы оставляем для сохранения обратной совместимости

L1feIsGood commented on LegacyApp/UserService.cs:
```diff
@@ -4,108 +4,63 @@ namespace LegacyApp
             }
 
-            if(age < 21)
+            if(!(mCreditLimitsService.SetCreditLimit(user)
+                                     .ValidateCreditLimit(user)))
```
Ах, вот зачем method chaining! 
Выглядит очень странно и вызывает кучу вопросов:
 - что SetCreditLimit делает в проверке условия?
 - ValidateCreditLimit обязательно вызывать после SetCreditLimit, или можно сразу? 
 - если я сразу вызову ValidateCreditLimit, то я получу другой результат?
 - где хранится и на что влияет результат работы SetCreditLimit? в user-е или mCreditLimitsService?

Nerewan replied:
1. так было удобно написать, непременной важности именно такой формат не имеет, но так было чуть короче без потери смысла
2. можно-то можно, но тогда он будет обрабатывать текущие данные, до их изменения SetCreditLimit'ом
3. да, но см.п.2
4. метод меняет данные в user'е, возврат this сделан ради method chaining

L1feIsGood replied:
Не хотелось ввязываться в полемику, но нужно донести мысль)
Вопросы были заданы чисто риторические, проблема в том, что само их наличие говорит о неудачном проектировании работы класса. Для ситуации в п.2 даже есть специальная метрика или правило:
Чем проще пользователю модуля использовать его неправильно, тем он хуже спроектирован. В идеале нужно проектировать классы так, чтобы возможности использовать их неправильно не существовало.

L1feIsGood commented on LegacyApp/UserService.cs:
```diff
@@ -4,108 +4,63 @@ namespace LegacyApp
-                user.HasCreditLimit = false;
-                return;
-            }
+            var clientRepository = new ClientRepository();
```
А почему ClientRepository остался в агрегации, а не выехал в конструктор?

Nerewan replied:
Просто забыл про него, увы (=.=)"

L1feIsGood reviewed (commented):
Удивительно, но первая попытка была удачней, хотя бы потому что не привнесла кучи сомнительных решений.
